### PR TITLE
WdaLocalPort change implementation 

### PR DESCRIPTION
**Pull request by cyildirim**
I noticed i couldn't change WDA port over command line in spite of that I wrote parameter to the capabilities. After that change, WDA port can be change and it is going to be possible run parallel test on a macOs.

I change a few thing over webdriver.js to give parameter to WebDriverAgent port via command line arg. But it requires WebDriverAgent changes. If I should open one more pull-request to WebDriverAgent repo please inform me. Or it can be add to https://github.com/appium/appium-xcuitest-driver/blob/master/docs/real-device-config.md .

WebDriver Agent side changes: 
image

### CODE REVIEW

**Comment by imurchie** [comment]
This capability is not intended to  change the port that WebDriverAgent listens on, but the port that `iproxy` listens on locally.

If you want to change the port that WDA is running on, on the device, that will need a separate capability, and a change to WDA, unless WDA is already capable of doing it.

**Comment by cyildirim** [comment]
In this case. It can be change both of them webdriveragent port and iproxy. And it  is able to parallel running when set these ports different. (For example 8200) I tested it and it works properly. 

Is It wrong anything doesn't change despite i change the WdaLocalPort ?

**Comment by imurchie** [comment]
I'd rather there be a new capability, since it would be a breaking change to have `wdaLocalPort` suddenly do something different.

I would suggest `wdaRemotePort`.

File: lib/webdriveragent.js [hunk 1]
```diff
@@ -192,6 +192,9 @@ class WebDriverAgent {
     const versionMatch = new RegExp(/^(\d+)\.(\d+)/).exec(this.platformVersion);
     if (versionMatch) {
       args.push(`IPHONEOS_DEPLOYMENT_TARGET=${versionMatch[1]}.${versionMatch[2]}`);
+      if (this.wdaLocalPort) {
+        args.push(" WDAPORT="+this.wdaLocalPort);
```
Comment (imurchie) [inline lib/webdriveragent.js]: Please use string interpolation: 
```
args.push(`WDAPORT=${this.wdaRemotePort}`);
```

**Comment by cyildirim** [comment]
Ok i will add it as new capability and i am going to push it again soon . Thanks

**Comment by cyildirim** [comment]
I pushed code and opened pull request facebook/WebDriverAgent side.

**Comment by imurchie** [comment]
https://github.com/facebook/WebDriverAgent/pull/573

**Comment by cyildirim** [comment]
It looks like everything is ok with the iproxy. Sorry and Thanks for your time @imurchie .

**Comment by NozomiIto** [comment]
Can we now use the port other than 8100 for the WDA process in XCUITestDriver for iOS simulator?
The latest XCUITestDriver still seems not to support this feature to me (https://github.com/appium/appium-xcuitest-driver/blob/master/lib/wda/webdriveragent.js#L161).

I'm very happy if I can use the port other than 8100 for WDA since 8100 port in my PC is unfortunately used by unknown process [x-print server](http://www.networksorcery.com/enp/protocol/ip/ports08000.htm) ..

**Comment by NozomiIto** [comment]
@cyildirim Thanks for your quick reply. Yes, I could also change the WDA port by wdaLocalPort option as for the real device. But as for the iOS simulator, this solution did not work..

**Comment by NozomiIto** [comment]
I tried changing [this line](https://github.com/appium/appium-xcuitest-driver/blob/master/lib/wda/webdriveragent.js#L158) from 
`if (this.realDevice && this.wdaLocalPort) {` 
to 
`if (this.wdaLocalPort) {` 
on my local PC and I found the log line like `[JSONWP Proxy] Proxying [GET /status] to [GET http://localhost:8200/status] with no body` were output to the log console also for iOS simulator, but I encountered the another following error.
```
[MJSONWP] Encountered internal error running command: Error: Unable to start WebDriverAgent session: Could not proxy command to remote server. Original error: Error: connect ECONNREFUSED 127.0.0.1:8200
test failed
```

**Comment by cyildirim** [comment]
Can you try with this change and also change on WebdriverAgent ? If it will work, it may be need to apply this change. 

https://github.com/facebook/WebDriverAgent/pull/573

**Comment by NozomiIto** [comment]
@cyildirim Thanks! I applied this change, but still I encountered `[MJSONWP] Encountered internal error running command: Error: Unable to start WebDriverAgent session: Could not proxy command to remote server. Original error: Error: connect ECONNREFUSED 127.0.0.1:8200
test failed`.

But, if your commits https://github.com/appium/appium-xcuitest-driver/pull/435/commits/4573001bd5d6b70317622331a5dce51a8ea18dce and https://github.com/appium/appium-xcuitest-driver/pull/435/commits/615d4efcbc24e780205cf494ffef8045ec58a75c are merged to master, I feel all problems will be resolved. ( are they still under review process?)

**Comment by imurchie** [comment]
Removing the real device check will _not_ work, since it just changes where Appium sends the requests, not where the server is actually running.

Right now this is not possible. We are working on a different way of starting the WebDriverAgent server, which will allow it to work.

**Comment by NozomiIto** [comment]
@imurchie Thank you! I understand, and am looking forward to the update!

**Comment by imurchie** [comment]
@NozomiIto Can you try with the latest beta (`npm uninstall -g appium && npm cache clean && npm install -g appium@beta`). It should have `appium-xcuitest-driver` version 2.32.0. Setting the `wdaLocalPort` with a simulator will load the server on the simulator running on the port specified.

**Comment by NozomiIto** [comment]
@imurchie Thanks for your quick response! 

## simulator

When I tried it on the iOS simulator with wdaLocalPort: 8110, it works well. Thanks you so much!

## real device

At the same time, I also tried this on the real device, and I encountered 

```
Encountered internal error running command: Error: Unable to launch WebDriverAgent because of xcodebuild failure: Cannot read property 'udid' of undefined
```

I added the stack trace logic to my local js file, and got the following stack trace:

```
TypeError: Cannot read property 'udid' of undefined
    at new iProxy (../../../lib/wda/iproxy.js:15:72)
    at WebDriverAgent.launch$ (../../../lib/wda/webdriveragent.js:101:21)
    ....
```

It seems the field `this.device` is undefined in [this line](https://github.com/appium/appium-xcuitest-driver/blob/4470b8949e024bddca738188274cc6d1e37a471b/lib/wda/iproxy.js#L15).

**Comment by imurchie** [comment]
@NozomiIto Can you post a link to a [gist](https://gist.github.com) of the _full_ Appium server logs, so I can see the configuration that causes this?

**Comment by imurchie** [comment]
Also: something with your configuration. This code should only be used when running on a real device.

**Comment by NozomiIto** [comment]
I uploaded  [here](https://gist.github.com/NozomiIto/4fd4db7ccfe69a85c1bfbc964597c04a).

- some udid and xcodeOrgId are masked
- I launched appium server without any log

**Comment by imurchie** [comment]
The `udid` you are using is that of a real device.

**Comment by NozomiIto** [comment]
Oh, yes. I tried your change both on the simulator and real device. Only on the real device, I encountered this problem.

**Comment by imurchie** [comment]
My change does not do anything for real devices. Real devices work as they did. I've pushed a fix for the issue, and updated the published version.

**Comment by NozomiIto** [comment]
Yeah, XCUITestDriver 2.31.0 worked on my environment both for the real device and simulator, and I also think wdaLocalPort fix causes no problem.

But 2.32.0 includes one more another commit https://github.com/appium/appium-xcuitest-driver/commit/4470b8949e024bddca738188274cc6d1e37a471b, 
and I think [this line](https://github.com/appium/appium-xcuitest-driver/blob/4470b8949e024bddca738188274cc6d1e37a471b/lib/wda/iproxy.js#L15) is added by this commit, and `this.device.udid` in this line causes the exception.

**Comment by imurchie** [comment]
Yes, that is fixed in 2.32.1.

**Comment by NozomiIto** [comment]
Thank you  😄  I have no time now, but I will try it later!

**Comment by NozomiIto** [comment]
2.32.2 worked on my environment. Thank you!

**Pull request closed without merging**